service/api_patch.go

```diff
@@ -35,36 +35,102 @@ type PatchAPIResponse struct {
 	Patch   *patch.Patch `json:"patch"`
 }
 
+// getAuthor returns the author for the patch. If githubAuthor or patchAuthor is provided and exists, will use this instead
+// of the submitter if the submitter is authorized to submit patches on behalf of users.
+// Returns the author, status code, and error.
+func (as *APIServer) getAuthor(data patchData, dbUser *user.DBUser, projectId, patchID string) (string, int, error) {
+	author := dbUser.Id
+	if data.GithubAuthor == "" && data.PatchAuthor == "" {
+		return author, http.StatusOK, nil
+	}
+
+	opts := gimlet.PermissionOpts{
+		Resource:      projectId,
+		ResourceType:  evergreen.ProjectResourceType,
+		Permission:    evergreen.PermissionPatches,
+		RequiredLevel: evergreen.PatchSubmitAdmin.Value,
+	}
+	if !dbUser.HasPermission(opts) {
+		return "", http.StatusUnauthorized, errors.New("user is not authorized to patch on behalf of other users")
+	}
+
+	if data.GithubAuthor != "" {
+		specifiedUser, err := user.FindByGithubName(data.GithubAuthor)
+		if err != nil {
+			return "", http.StatusInternalServerError, errors.Wrap(err, "error looking for specified author")
+		}
+		if specifiedUser != nil {
+			grip.Info(message.Fields{
+				"message":               "overriding patch author as specified by the submitter",
+				"submitter":             dbUser.Id,
+				"new_author":            specifiedUser.Id,
+				"given_github_username": data.GithubAuthor,
+				"patch_id":              patchID,
+			})
+			author = specifiedUser.Id
+		}
+		grip.DebugWhen(specifiedUser == nil, message.Fields{
+			"message":         "github user not found",
+			"github_username": data.GithubAuthor,
+			"patch_id":        patchID,
+		})
+	} else if data.PatchAuthor != "" {
+		specifiedUser, err := user.FindOneById(data.PatchAuthor)
+		if err != nil {
+			return "", http.StatusInternalServerError, errors.Wrap(err, "error looking for specified author")
+		}
+		if specifiedUser != nil {
+			grip.Info(message.Fields{
+				"message":    "overriding patch author as specified by the submitter",
+				"submitter":  dbUser.Id,
+				"new_author": data.PatchAuthor,
+				"patch_id":   patchID,
+			})
+			author = specifiedUser.Id
+		}
+		grip.DebugWhen(specifiedUser == nil, message.Fields{
```

Comment:
If someone uses a nonexistent author name or has a typo, can we propagate down to the CLI output that the user was not found? It seems that with the exception of this debug statement it would fail silently and the author name would default to the submitter's username.

Reply:
The behavior of this command for the commit queue and the desired behavior for this change is that it would default to the author name if the submitter name isn't configured. This behavior is in the flag description https://github.com/evergreen-ci/evergreen/blob/4afe0f066dd383e47d54cac87858ba7d3d417f11/operations/patch.go#L290

Reply:
Does the CLI output mention if the user was found or not if an author name is provided? Or is the only way for us to find out if the author override was successful to check the submitter field in the UI for the patch?

My only concern here was that if there isn't any direct CLI feedback that a default / fallback to the submitter is taking place.

Reply:
Does the CLI output mention if the user was found successfully if an author name is provided? Or is the only way for us to find out if the author override was successful to check the submitter field in the UI for the patch?

My only concern here was that if there isn't any direct CLI feedback that a default / fallback to the submitter is taking place.

Reply:
Does the CLI output mention if the user was found successfully if an author name is provided? Or is the only way for us to find out if the author override was successful to check the submitter field in the UI for the patch?

My only concern here was that if there isn't any direct CLI feedback that a default / fallback to the submitter is taking place.

Reply:
The CLI output doesn't mention or know about whether it's succeeded yet, bc this logic is happening on the app and not on the client. I'm still fine with that since this permission is only given to bots (i.e. this permission isn't grantable on mana or given to project admins or superusers) that probably aren't checking if it ended up defaulting anyway.